zipkin-cassandra/src/main/scala/com/twitter/zipkin/storage/cassandra/CassandraSpanStore.scala

```diff
@@ -277,71 +220,48 @@ class CassieSpanStore(
+    FuturePool.unboundedPool { repository.close() }
 
-  // TODO: break these into smaller batches?
   def apply(spans: Seq[Span]): Future[Unit] = {
     SpansStoredCounter.incr(spans.size)
 
-    val traces = Traces.batch()
-    val serviceNames = ServiceNames.batch()
-    val spanNames = SpanNames.batch()
-    val serviceNameIdx = ServiceNameIndex.batch()
-    val serviceSpanNameIdx = ServiceSpanNameIndex.batch()
-    val annotationsIdx = AnnotationsIndex.batch()
-    val durationIdx = DurationIndex.batch()
-
     spans foreach { span =>
-      traces.insert(span.traceId, Column[String, ThriftSpan](createSpanColumnName(span), span.toThrift, None, Some(spanTtl)))
+      repository.storeSpan(span.traceId, createSpanColumnName(span), spanCodec.encode(span.toThrift), spanTtl.inSeconds)
       if (shouldIndex(span)) {
         SpansIndexedCounter.incr()
-        indexServiceName(serviceNames, span)
-        indexSpanNameByService(spanNames, span)
-        indexTraceIdByName(serviceNameIdx, serviceSpanNameIdx, span)
-        indexByAnnotations(annotationsIdx, span)
-        indexSpanDuration(durationIdx, span)
+        indexServiceName(span)
+        indexSpanNameByService(span)
+        indexTraceIdByName(span)
+        indexByAnnotations(span)
+        indexSpanDuration(span)
       }
     }
 
-    Future.collect(Seq(
-      traces,
-      serviceNames,
-      spanNames,
-      serviceNameIdx,
-      serviceSpanNameIdx,
-      annotationsIdx,
-      durationIdx
-    ).map(_.execute())).unit
+    Future.Unit
   }
 
   def setTimeToLive(traceId: Long, ttl: Duration): Future[Unit] = {
-    Traces.getRow(traceId) flatMap { row =>
-      val traces = Traces.batch()
-      row.values.asScala foreach { col =>
-        traces.insert(traceId, col.copy(timestamp = None, ttl = Some(ttl)))
-      }
-      traces.execute().unit
+    getSpansByTraceId(traceId).get foreach { span =>
+      repository.storeSpan(traceId, createSpanColumnName(span), spanCodec.encode(span.toThrift), ttl.inSeconds)
     }
+    Future.Unit
   }
 
   def getTimeToLive(traceId: Long): Future[Duration] = {
     QueryGetTtlCounter.incr()
-    Traces.getRow(traceId) flatMap { row =>
-      val ttl = row.values.asScala.foldLeft(Int.MaxValue) { (ttl, col) =>
-        math.min(ttl, col.ttl.map(_.inSeconds).getOrElse(Int.MaxValue))
-      }
 
-      if (ttl == Int.MaxValue)
-        Future.exception(new IllegalArgumentException("The trace " + traceId + " does not have any ttl set!"))
-      else
-        Future.value(ttl.seconds)
+    pool {
+      Duration(repository.getSpanTtl(traceId), java.util.concurrent.TimeUnit.SECONDS)
     }
   }
 
   def tracesExist(traceIds: Seq[Long]): Future[Set[Long]] = {
     QueryTracesExistStat.add(traceIds.size)
-    getSpansByTraceIds(traceIds, 1) map {
-      _.flatMap(_.headOption.map(_.traceId)).toSet
+    pool {
+      repository
+        .tracesExist(traceIds.toArray.map(Long.box))
+        .map(_.asInstanceOf[Long])
```

Comment:
I think this is the same as `.map(Long.unbox)` (mentioning it since `Long.box` was used above)